pkg/gadget-context/run.go

```diff
@@ -0,0 +1,162 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gadgetcontext
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	apihelpers "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api-helpers"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+)
+
+func (c *GadgetContext) initAndPrepareOperators(paramValues api.ParamValues) ([]operators.DataOperatorInstance, error) {
+	log := c.Logger()
+
+	ops := c.DataOperators()
+
+	// Sort dataOperators based on their priority
+	sort.Slice(ops, func(i, j int) bool {
+		return ops[i].Priority() < ops[j].Priority()
+	})
+
+	params := make([]*api.Param, 0)
+
+	dataOperatorInstances := make([]operators.DataOperatorInstance, 0, len(ops))
+	for _, op := range ops {
+		log.Debugf("initializing data op %q", op.Name())
+		opParamPrefix := fmt.Sprintf("operator.%s", op.Name())
+
+		// Lazily initialize operator
+		// TODO: global params should be filled out from a config file or such; maybe it's a better idea not to
+		// lazily initialize operators at all, but just hand over the config. The "lazy" stuff could then be done
+		// if the operator is instantiated and needs to do work
+		err := op.Init(apihelpers.ToParamDescs(op.GlobalParams()).ToParams())
+		if err != nil {
+			return nil, fmt.Errorf("initializing operator %q: %w", op.Name(), err)
+		}
+
+		// Get and fill params
+		instanceParams := op.InstanceParams().AddPrefix(opParamPrefix)
+		opParamValues := paramValues.ExtractPrefixedValues(opParamPrefix)
+
+		params = append(params, instanceParams...)
+
+		err = apihelpers.Validate(instanceParams, opParamValues)
+		if err != nil {
+			return nil, fmt.Errorf("validating params for operator %q: %w", op.Name(), err)
+		}
+
+		opInst, err := op.InstantiateDataOperator(c, opParamValues)
+		if err != nil {
+			log.Errorf("instantiating operator %q: %v", op.Name(), err)
+		}
+		if opInst == nil {
+			log.Debugf("> skipped %s", op.Name())
+			continue
+		}
+		dataOperatorInstances = append(dataOperatorInstances, opInst)
+	}
+
+	for _, opInst := range dataOperatorInstances {
+		log.Debugf("preparing op %q", opInst.Name())
+		opParamPrefix := fmt.Sprintf("operator.%s", opInst.Name())
+
+		// Second pass params; this time the operator had the chance to prepare itself based on DataSources, etc.
+		// this mainly is postponed to read default values that might differ from before; this second pass is
+		// what is handed over to the remote end
+		if extra, ok := opInst.(operators.DataOperatorExtraParams); ok {
+			pd := extra.ExtraParams(c)
+			params = append(params, pd.AddPrefix(opParamPrefix)...)
+		}
```

Comment:
It's important to have

```
op1.Init()
op2.Init()
op1.ExtraParams()
op2.ExtraParams()
``

instead of 

```
op1.Init()
op1.ExtraParams()
op2.Init()
op2.ExtraParams()
``

Because the ExtraParams in this example depend on all operators being initialized.

Reply:
The `ExtraParams()` could actually work in the loop above, but I'd also consider them a lifecycle hook and I want to keep them cleanly separated. IMHO, there has to be a guarantee that all operators have run the previous lifecycle method before continuing to the next.